src/us/kbase/workspace/kbase/DelegatingTypeServerMethods.java

```diff
@@ -0,0 +1,158 @@
+package us.kbase.workspace.kbase;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+
+import us.kbase.auth.AuthToken;
+import us.kbase.workspace.FuncInfo;
+import us.kbase.workspace.GetModuleInfoParams;
+import us.kbase.workspace.GrantModuleOwnershipParams;
+import us.kbase.workspace.ListAllTypesParams;
+import us.kbase.workspace.ListModuleVersionsParams;
+import us.kbase.workspace.ListModulesParams;
+import us.kbase.workspace.ModuleInfo;
+import us.kbase.workspace.ModuleVersions;
+import us.kbase.workspace.RegisterTypespecCopyParams;
+import us.kbase.workspace.RegisterTypespecParams;
+import us.kbase.workspace.RemoveModuleOwnershipParams;
+import us.kbase.workspace.TypeInfo;
+import us.kbase.workspace.kbase.WorkspaceDelegator.WorkspaceDelegationException;
+
+/** An implementation of the type service methods that delegate to another workspace service.
+ * Note that any asAdmin param toggles are ignored - administration methods should be delegated
+ * to the administration interface, not here. 
+ */
+public class DelegatingTypeServerMethods implements TypeServerMethods {
+	
+	private final WorkspaceDelegator delegator;
+	
+	/** Create the type service methods delegator.
+	 * @param delegator a workspace method delegator.
+	 */
+	public DelegatingTypeServerMethods(final WorkspaceDelegator delegator) {
+		this.delegator = requireNonNull(delegator, "delegator");
+	}
+	
+	@Override
+	public void grantModuleOwnership(
+			final GrantModuleOwnershipParams params,
+			final AuthToken token,
+			final boolean asAdmin) // admin methods should not call this when delegating
+			throws WorkspaceDelegationException {
+		delegator.delegate(token, c -> {c.grantModuleOwnership(params); return null;});
+	}
+
+	@Override
+	public void removeModuleOwnership(
+			final RemoveModuleOwnershipParams params,
+			final AuthToken token,
+			final boolean asAdmin) // admin methods should not call this when delegating
+			throws WorkspaceDelegationException {
+		delegator.delegate(token, c -> {c.removeModuleOwnership(params); return null;});
+	}
+
+	@Override
+	public void requestModuleOwnership(final String mod, final AuthToken token)
+			throws WorkspaceDelegationException {
+		delegator.delegate(token, c -> {c.requestModuleOwnership(mod); return null;});
```

Comment:
Why do these specifically return `null`? Shouldn't they naturally return `null` if structured as below, since the delegated method returns `null`? Or would they return some kind of empty container?

Reply:
I had a lot of fun figuring this out. I'm not totally sure I 100% grok it but this is my understanding of what's going on.

The `delegate` method takes a `WorkspaceCommand<T>`, which is a functional interface with a parameterized return type. That means any lambda functions passed as a `WorkspaceCommand` *must* return something. The 3 methods here that are suffixed with an explicit return statement all return `void`, which means they return nothing - not even `null`. If you take out those return statements you get a compiler error complaining that it can't figure out the type of the return value of the lambda function.

TLDR - if you take out the explicit return the code doesn't compile.

Reply:
I get it. That's annoying, but I get it.